File: DanBonehClass/VideoSHA.pl
use Digest::SHA;
use BlockRead;
use Utils;


die("Bad Command Line\n") unless @ARGV == 1;
my $fname = $ARGV[0];



my @blocks = BlockRead::BlockRead($fname,1024);

#print Utils::bintohex($blocks[0]),"\n";
#print Utils::bintohex($blocks[@blocks-1]),"\n";

my $state = Digest::SHA->new(256);
for (my $i = 0 ; $i < @blocks ; ++$i)
{
	$state->add($blocks[$i]);
}

print Utils::bintohex($state->digest),"\n";

$state->reset();
$state->addfile($fname);
print Utils::bintohex($state->digest),"\n";
exit(1);


for (my $i = ((scalar @blocks) - 1) ; $i >= 1 ; --$i )
{
	my $digest = Digest::SHA::sha256($blocks[$i]);
#	print "pre: ",length($blocks[$i-1]);
	$blocks[$i-1] .= $digest;
#	print " post: ",length($blocks[$i-1]),"\n";
}

print Utils::bintohex(Digest::SHA::sha256($blocks[0])),"\n";

File: DanBonehClass/VideoSHA.java
import java.io.File;
import java.io.FileInputStream;
import java.security.MessageDigest;


public class VideoSHA
{

	

	

	
	
	
	public static byte[] sha256(byte[] bytes)
	{
		byte[] result = null;
		
		try
		{
			MessageDigest digest = MessageDigest.getInstance("SHA-256");
			result = digest.digest(bytes);
		} catch (Exception ex)
		{
			System.out.println("Digesting Message: " + ex);
			System.exit(1);
		}				
		return result;
	}
	
	public static void main(String args[])
	{
		if (args.length != 1)
		{
			System.out.println("Bad Command Line");
			System.exit(1);
		}
		String fname = args[0];
		byte[] contents = null;
		
		try
		{
			File file=new File(fname);
			long size=file.length();
			contents=new byte[(int)size];
			FileInputStream in=new FileInputStream(fname);
			in.read(contents);
			in.close();
		} catch (Exception ex)
		{
			System.out.println("Exception caught on load: " + ex);
			System.exit(1);
		}
		
		byte[][] blocks = Blockify(contents,1024);
		for (int i = blocks.length - 1; i >= 1 ; --i)
		{
			blocks[i-1] = append(blocks[i-1],sha256(blocks[i]));
		}
		
		System.out.println("SHA-256 hash: " + BytesToHex(sha256(blocks[0])));
		
	}
}
